mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp

```diff
@@ -161,8 +193,7 @@ bool LLVM::StoreOp::canUsesBeRemoved(
   // fine, provided we are currently promoting its target value. Don't allow a
   // store OF the slot pointer, only INTO the slot pointer.
   return blockingUse == slot.ptr && getAddr() == slot.ptr &&
-         getValue() != slot.ptr && getValue().getType() == slot.elemType &&
-         !getVolatile_();
+         getValue() != slot.ptr && !getVolatile_();
```

Comment:
Should there be a check for cast compatibility here as well? I assume the conversion in `getStored` has to succeed?

Reply:
Nice catch. Working with a additional changes in the same files is a bad idea...
At least this shows that we should add a test for this case to ensure that this is detected.